prereise/gather/demanddata/transportation_electrification/smart_charging.py

```diff
@@ -117,6 +123,90 @@ def smart_charging(
     else:
         charging_efficiency = 0.9
 
+    # Add booleans for whether the location allows charging
+    if location_strategy == 3:
+        newdata["location_allowed"] = True
+    else:
+        if veh_type.lower() in {"ldv", "ldt"}:
+            allowed = const.ldv_location_allowed[location_strategy]
+            newdata["location_allowed"] = newdata["dwell_location"].isin(allowed)
+        elif veh_type.lower() in {"mdv", "hdv"}:
+            allowed = const.hdv_location_allowed[location_strategy]
+            newdata["location_allowed"] = newdata["dwell_location"].isin(allowed)
+
+    # Add booleans for whether the trip_number (compared to total trips) allows charging
+    if trip_strategy == 1:
+        newdata["trip_allowed"] = True
+    elif trip_strategy == 2:
+        newdata["trip_allowed"] = newdata["trip_number"] == newdata["total_trips"]
+
+    # Add booleans for whether the dwell time is long enough to allow charging
+    newdata["dwell_allowed"] = newdata["dwell_time"] > 0.2
+
+    # Add boolean for whether this trip allows charging
+    allowed_cols = [
+        "location_allowed",
+        "trip_allowed",
+        "dwell_allowed",
+    ]
+    newdata["charging_allowed"] = newdata[allowed_cols].apply(all, axis=1)
+
+    newdata["dwell_charging"] = (
+        newdata["charging_allowed"]
+        * newdata["dwell_time"]
+        * power
+        * charging_efficiency
+    )
+
+    grouped_trips = newdata.groupby("vehicle_number")
+    for vehicle_num, group in grouped_trips:
+        newdata.loc[group.index, "max_charging"] = newdata.loc[
+            group.index, "dwell_charging"
+        ].sum()
+        newdata.loc[group.index, "required_charging"] = (
+            newdata.loc[group.index, "trip_miles"].sum() * kwhmi
+        )
```

Comment:
Similar refactor suggestions as above.

Reply:
Same reply from me from `immediate.py`
In an ideal world, these filtering steps would be included in a separate function before `immediate` and `smart` charging are called.  That's a bonus refactor for later if there's time.